pyro/infer/autoguide/guides.py

```diff
@@ -347,6 +347,145 @@ def median(self, *args, **kwargs):
         return self(*args, **kwargs)
 
 
+class AutoNormal(AutoGuide):
+    """This is AutoNormal, but it uses the same shape constraint logic found
+    in AutoMultivariateNormal and AutoDiagonalNormal.
+    """
+    def __init__(self, model, init_loc_fn=init_to_median, prefix='auto'):
+        #  Does this accept a 'prefix' arg?  It doesn't look like it.
+        self.init_loc_fn = init_loc_fn
+        self.prefix = prefix
+        model = InitMessenger(self.init_loc_fn)(model)
+        super().__init__(model)
+
+    def _setup_prototype(self, *args, **kwargs):
+        super()._setup_prototype(*args, **kwargs)
+
+        self._unconstrained_shapes = {}
+        self._cond_indep_stacks = {}
+
+        # Initialize guide params
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            # Collect the shapes of unconstrained values.
+            # These may differ from the shapes of constrained values.
+            constrained_shape = site["value"].shape
+            unconstrained_shape = biject_to(site["fn"].support).inv(site["value"]).shape
+
+            event_dim = site["fn"].event_dim + len(unconstrained_shape) - len(constrained_shape)
+
+            site_batch_shape = site["fn"].batch_shape
+
+            self._unconstrained_shapes[name] = unconstrained_shape
+            unconstrained_shape_for_params = broadcast_shape(
+                unconstrained_shape,
+                site_batch_shape + (1,) * event_dim
+            )
+
+            # Collect independence contexts.
+            self._cond_indep_stacks[name] = site["cond_indep_stack"]
+
+            #  Set up params
+            loc_name = "{}_{}_{}".format(self.prefix, name, 'loc')
+            scale_name = "{}_{}_{}".format(self.prefix, name, 'scale')
+
+            pyro.param(
+                loc_name,
+                lambda: site["value"].new_zeros(unconstrained_shape_for_params)
+            )
+            pyro.param(
+                scale_name,
+                lambda: site["value"].new_ones(unconstrained_shape_for_params),
+                constraint=constraints.positive
+            )
+
+    def _get_loc_and_scale(self, name):
+        loc_name = "{}_{}_{}".format(self.prefix, name, 'loc')
+        scale_name = "{}_{}_{}".format(self.prefix, name, 'scale')
+        site_loc = pyro.get_param_store().get_param(loc_name)
+        site_scale = pyro.get_param_store().get_param(scale_name)
```

Comment:
...then you can update this block to
```py
site_loc = getattr(self.locs, name)
site_scale = getattr(self.scales, name)
```

Reply:
So I made those changes - and now the serialization test passes (🎉 ) - but I'm getting `E           KeyError: 'parameter name can\'t contain "."'` errors from the register_parameter method in the test_predictive() test, which used to work.

The names in get_param_store() are now `AutoNormal.locs.<site_name>`.  Is there a way to control these?  Using `self.locs = PyroModule(name='locs')` appears not to change anything.

Reply:
Ah - I commented too soon.  Looks like the PyroModule() edits also make the median() and quantiles() tests fail (`AssertionError: cannot compare <class 'torch.nn.parameter.Parameter'> and <class 'torch.Tensor'>).  I try to get them working again tomorrow 👍 .

Reply:
> Is there a way to control [names in the param store]?

Pyro 1.0 gives up some naming freedom in return for serializability and servability (Pyro models can now be served even in C++!). That said, we could change design a bit to allow a different standard naming scheme. I suggested `AutoNormal.locs.<site_name>` because it appeared to minimize risk of name conflict with `dir(AutoNormal)` (which could happen if e.g. a user named a site `median` or `init_scale`). Did you have a different naming convention in mind?